crates/viewer/re_viewer_context/src/drag_and_drop.rs

```diff
@@ -0,0 +1,184 @@
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Contents { contents } => items_to_string(
+                contents
+                    .iter()
+                    .map(|content| content.as_item())
+                    .collect_vec()
+                    .iter(),
+            )
+            .fmt(f),
+
+            // this is not used in the UI
+            Self::Invalid => "invalid selection".fmt(f),
+        }
+    }
+}
+
+/// Display the currently dragged payload as a pill in the UI.
+///
+/// This should be called once per frame.
+pub fn drag_and_drop_payload_cursor_ui(ui: &mut egui::Ui) {
+    if let Some(payload) = egui::DragAndDrop::payload::<DragAndDropPayload>(ui.ctx()) {
+        if let Some(pointer_pos) = ui.ctx().pointer_interact_pos() {
+            let icon = match payload.as_ref() {
+                DragAndDropPayload::Contents { .. } => &re_ui::icons::DND_MOVE,
+                // don't draw anything for invalid selection
+                DragAndDropPayload::Invalid => return,
+            };
+
+            let layer_id = egui::LayerId::new(
+                egui::Order::Tooltip,
+                egui::Id::new("drag_and_drop_payload_layer"),
+            );
+            let response = ui
+                .scope_builder(egui::UiBuilder::new().layer_id(layer_id), |ui| {
+                    ui.set_opacity(0.7);
+
+                    drag_pill_frame(matches!(
+                        payload.as_ref(),
+                        &DragAndDropPayload::Invalid { .. }
+                    ))
+                    .show(ui, |ui| {
+                        let text_color = ui.visuals().widgets.inactive.text_color();
+
+                        ui.horizontal(|ui| {
+                            ui.spacing_mut().item_spacing.x = 2.0;
+
+                            ui.small_icon(icon, Some(text_color));
+                            ui.label(egui::RichText::new(payload.to_string()).color(text_color));
+                        });
+                    })
+                    .response
+                })
+                .response;
+
+            let delta = pointer_pos - response.rect.right_bottom();
+            ui.ctx()
+                .transform_layer_shapes(layer_id, emath::TSTransform::from_translation(delta));
+        }
+    }
+}
+
+fn drag_pill_frame(error_state: bool) -> egui::Frame {
+    let hue = if error_state { Hue::Red } else { Hue::Blue };
+
+    egui::Frame {
+        fill: re_ui::design_tokens().color(ColorToken::new(hue, S325)),
+        stroke: egui::Stroke::new(
+            1.0,
+            re_ui::design_tokens().color(ColorToken::new(hue, S375)),
+        ),
+        rounding: (2.0).into(),
+        inner_margin: egui::Margin {
+            left: 6.0,
+            right: 9.0,
+            top: 5.0,
+            bottom: 4.0,
+        },
+        ..Default::default()
+    }
+}
+
+fn items_to_string<'a>(items: impl Iterator<Item = &'a Item>) -> String {
```

Comment:
When dragging a single item (e.g. a space-view), would it be nice to show its name instead? @gavrelina

Reply:
I'll reject this for complexity over value reason. That would require the payload stuff to query the blueprint store to retrieve the name.

If you think it's important, please open an issue.